### ISSUE TITLE
Lazy seqs should be realized before handing segment to the downstream task

### ISSUE THREAD

**Issue by jetmind** [onyx-platform/onyx#755]
It seems like lazy sequences created in one task can be realized in the following tasks. This causes problems when using flow conditions with `:flow/thrown-exception? true` and relying on the task name for specific error handling in the post-transform hooks. I.e. exception is actually thrown in the downstream task but logically this error is in upstream task and should be handled as if it were happened there.

Here's the repo with minimal reproduction https://github.com/jetmind/onyx-lazy-error. Just run `lein test` on it and see comments in the code. Important parts are [this](https://github.com/jetmind/onyx-lazy-error/blob/master/src/onyx_lazy_error/core.clj#L90) and [this](https://github.com/jetmind/onyx-lazy-error/blob/master/test/onyx_lazy_error/core_test.clj#L62-L68).

Please see original slack thread for more context:
<details>
<summary>Slack log</summary>
jetmind [17:27] 
Hey folks! I have a weird question about lazy sequences processing in onyx.

Let’s imagine I have a task which returns a segment containing lazy seq:

```{:numbers (for [i (range 10)] i)}```

and the downstream task which uses this lazy seq:

```(reduce + (:numbers segment))```

Would the lazy seq be realised before handing the segment to a task downstream or when it’s actually used in the downstream task? Maybe both variants are possible depending on the assignment of tasks to peers?

michaeldrogalis [17:34] 
@jetmind The segment is sent over the network to the next peer, probably on another machine, so it’s realized.

jetmind [17:50] 
That’s what I thought. The reason I asked is because I have a flow condition which reports exception and a task name it appeared in to sentry and I see a downstraem task name there, but stack trace indicates error inside the lazy seq in the upstream task.

michaeldrogalis [18:10] 
@jetmind Can you rephrase that? I’m a bit confused to what you’re describing.

jetmind [18:29] 
@michaeldrogalis Basically what I’m saying is that our error tracking system shows stack trace which indicates the error happened while realizing a lazy sequence (which is created in task1) and it seems as if this seq is realized in the task that follows the task1 and not in the task1 itself.

I get the name of the task I’m sending to the error tracking system when exception happens from `(-> event :onyx.core/task-map :onyx/name)` in the post-transform hook.

michaeldrogalis [18:30] 
@jetmind There’s one thing that _might_ make this possible, but I’d have to check.

[18:30] 
We’re using Nippy to serialize some types of data between networked machines

[18:31] 
It wouldn’t surprise me if Nippy had some tolerance for freezing/thawing lazy seqs.

jetmind [18:31] 
ok, I feel less insane now)

michaeldrogalis [18:31] 
I know that sounds weird, but I also recently learned that it propagates metadata too. Heh.

[18:31] 
Haha, yeah we’ll figure it out

jetmind [18:31] 
cool, thanks for your help!

michaeldrogalis [18:32] 
Wow, it does

[18:32] 
 ``` :lazy-seq       (repeatedly 1000 rand)
 :lazy-seq-empty (map identity '())
```

[18:32] 
That’s in it’s example README.

jetmind [18:41] 
As far as I understand it’s still realizing the seq before serialization, there’s no magic, right?

```(def frozen (nippy/freeze {:lazy (for [i (range 3)] (do (println i) i))}))
0
1
2
```

So I’m not sure I understand why realization appears as if it is done from the downstream task

michaeldrogalis [18:43] 
I don’t have time to dig deeper, but I can try to help some more later. That’s certainly odd behavior.

jetmind [18:46] 
Sure, thanks! I think I’ll just realize the seq manually in the first task, but it would be great to understand why this is happening in the first place.

[18:47] 
I depend on the task name to get some more data to report feedback to related systems and this behavior breaks things

michaeldrogalis [18:48] 
Totally makes sense. If possible, can you open an issue with a minimal reproduction?

[18:48] 
I know that’s a bit harder with these sorts of system, but if you can trim it down at all I’d greatly appreciate it.

jetmind [18:50] 
Sure, I’ll try to trim it down later

michaeldrogalis [18:57] 
Thanks :slightly_smiling_face:
</details>

**Comment by MichaelDrogalis** [comment]
Can you verify this against Onyx 0.10.0-beta12? This shouldn't be an issue there.

**Comment by jetmind** [comment]
Yeah, it works as expected with 0.10.0-beta12

**Comment by MichaelDrogalis** [comment]
Thanks for testing this. :)
